btcec/signature_test.go

```diff
@@ -491,3 +506,88 @@ func TestSignCompact(t *testing.T) {
 		testSignCompact(t, name, btcec.S256(), data, compressed)
 	}
 }
+
+func verifyRFC6979Nonce(hexKey string, msg string, hexNonce string, t *testing.T) {
+	pkBytes := decodeHex(hexKey)
+	nonce := decodeHex(hexNonce)
+	hash := fastsha256.Sum256([]byte(msg))
+	actualNonceBigInt := btcec.TstNonceRFC6979(pkBytes, hash[:])
+	actualNonce := actualNonceBigInt.Bytes()
+	if !bytes.Equal(actualNonce, nonce) {
+		t.Errorf("Nonce for message '%s' is incorrect: %s (expected %s)", msg, hex.EncodeToString(actualNonce), hex.EncodeToString(nonce))
+	}
+}
+
+func verifyRFC6979Signature(hexKey string, msg string, hexSignature string, t *testing.T) {
+	pkBytes := decodeHex(hexKey)
+	privKey, _ := btcec.PrivKeyFromBytes(btcec.S256(), pkBytes)
+	sigBytes := decodeHex(hexSignature)
+	hash := fastsha256.Sum256([]byte(msg))
+	actualSig, err := privKey.Sign(hash[:])
+	if err != nil {
+		panic("failed to generate signature in test source: err " + err.Error() +
+			", private key: " + hexKey + ", message: " + msg)
+	}
+	actualSigBytes := actualSig.Serialize()
+	if !bytes.Equal(actualSigBytes, sigBytes) {
+		t.Errorf("Signature for message '%s' is incorrect: %s (expected %s)", msg, hex.EncodeToString(actualSigBytes), hex.EncodeToString(sigBytes))
+	}
+}
+
+func TestRFC6979(t *testing.T) {
+	// Test vectors matching Trezor and CoreBitcoin implementations.
+	// - https://github.com/trezor/trezor-crypto/blob/9fea8f8ab377dc514e40c6fd1f7c89a74c1d8dc6/tests.c#L432-L453
+	// - https://github.com/oleganza/CoreBitcoin/blob/e93dd71207861b5bf044415db5fa72405e7d8fbc/CoreBitcoin/BTCKey%2BTests.m#L23-L49
+	verifyRFC6979Nonce(
```

Comment:
We've been trying to move everything over to table driven tests to be more consistent in the code.  Would you mind updating the test code to the following (after adding all the vectors in).  Note that I also simplified a couple of things such as using `%x` in the errors instead of manually encoding the bytes.  Also, I changed the error check on sign to a `t.Errorf` instead of panic since that isn't a hard coded value.

With this code the `verifyRFC6979Nonce` and `verifyRFC6979Signature` functions aren't needed.

```Go
func TestRFC6979(t *testing.T) {
	// Test vectors matching Trezor and CoreBitcoin implementations.
	// - https://github.com/trezor/trezor-crypto/blob/9fea8f8ab377dc514e40c6fd1f7c89a74c1d8dc6/tests.c#L432-L453
	// - https://github.com/oleganza/CoreBitcoin/blob/e93dd71207861b5bf044415db5fa72405e7d8fbc/CoreBitcoin/BTCKey%2BTests.m#L23-L49
	tests := []struct {
		key       string
		msg       string
		nonce     string
		signature string
	}{
		{
			"cca9fbcc1b41e5a95d369eaa6ddcff73b61a4efaa279cfc6567e8daa39cbaf50",
			"sample",
			"2df40ca70e639d89528a6b670d9d48d9165fdc0febc0974056bdce192b8e16a3",
			"3045022100af340daf02cc15c8d5d08d7735dfe6b98a474ed373bdb5fbecf7571be52b384202205009fb27f37034a9b24b707b7c6b79ca23ddef9e25f7282e8a797efe53a8f124",
		},
		// The rest....
	}

	for i, test := range tests {
		pkBytes := decodeHex(test.key)
		hash := fastsha256.Sum256([]byte(test.msg))

		// Ensure deterministically generated nonce is the expected
		// value.
		gotNonce := btcec.TstNonceRFC6979(pkBytes, hash[:]).Bytes()
		wantNonce := decodeHex(test.nonce)
		if !bytes.Equal(gotNonce, wantNonce) {
			t.Errorf("NonceRFC6979 #%d (%s): Nonce is incorrect: "+
				"%x (expected %x)", i, test.msg, gotNonce,
				wantNonce)
			continue
		}

		// Ensure deterministically generated signature is the expected
		// value.
		privKey, _ := btcec.PrivKeyFromBytes(btcec.S256(), pkBytes)
		gotSig, err := privKey.Sign(hash[:])
		if err != nil {
			t.Errorf("Sign #%d (%s): unexpected error: %v", i,
				test.msg, err)
			continue
		}
		gotSigBytes := gotSig.Serialize()
		wantSigBytes := decodeHex(test.signature)
		if !bytes.Equal(gotSigBytes, wantSigBytes) {
			t.Errorf("Sign #%d (%s): mismatched signature: %x "+
				"(expected %x)", i, test.msg, gotSigBytes,
				wantSigBytes)
			continue
		}
	}
}
```